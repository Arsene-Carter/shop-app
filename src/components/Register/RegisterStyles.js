import styled from 'styled-components';

export const RegisterContainer = styled.div`
  display: flex;
  width: 30%;
  flexDirection: column;
`;

export const RegisterTitle = styled.h2`
 flex: 1;
 alignSelf: center;
`;
