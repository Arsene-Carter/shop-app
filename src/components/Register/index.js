// @flow
import React, { Component } from 'react';
import { SessionContainer } from '../styles';
import { Title, Button, Input, Link, FormField } from '../../ui';
import * as validationErrors from '../ValidationErrors';
import FormWithError from '../FormComponents/FormError';

type RegisterFields = {
  email: string,
  password: string,
  confirmPassword: string
};

type TouchedState = {
  email: boolean,
  password: boolean,
  confirmPassword: boolean
};

type State = {
  email: string,
  password: string,
  confirmPassword: string,
  touchedFields: TouchedState
};

type Props = {
  register: ({
    email: string,
    password: string,
    confirmPassword: string
  }) => void
};

export const validate = ({
  email,
  password,
  confirmPassword,
  touchedFields
}: RegisterFields & { touchedFields: TouchedState }): {
  email: ?string,
  password: ?string,
  confirmPassword: ?string
} => {
  const validateRequired = (field: string) =>
    field ? '' : validationErrors.REQUIRED;
  const validateEmail = (emailStr: string) => {
    if (
      !emailStr.match(
        /^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/
      )
    ) {
      return validationErrors.INVALID_EMAIL;
    }
    return null;
  };

  const minChars = len => str =>
    str.length < len ? validationErrors.MIN_6_CHARACTERS : '';
  const min6 = minChars(6);
  const isTouched = field => touchedFields[field];
  const passwordMatch = () =>
    password === confirmPassword ? '' : validationErrors.PASSWORD_NOT_MATCH;

  return {
    email:
      (isTouched('email') &&
        (validateRequired(email) || validateEmail(email))) ||
        '',
    password:
      (isTouched('password') &&
        (validateRequired(password) || min6(password))) ||
        '',
    confirmPassword:
      (isTouched('confirmPassword') &&
        (validateRequired(confirmPassword) ||
          min6(password) ||
          passwordMatch())) ||
        ''
  };
};

const WithErrorInput = FormWithError(Input);

export default class Register extends Component<void, Props, State> {
  state = {
    email: '',
    password: '',
    confirmPassword: '',
    touchedFields: {
      email: false,
      password: false,
      confirmPassword: false
    }
  };

  inputHandler = (name: string) => (event: SyntheticInputEvent) =>
    this.setState({ [name]: event.target.value });
  handleOnBlur = (name: string) =>
    this.setState({
      touchedFields: Object.assign({}, this.state.touchedFields, {
        [name]: true
      })
    });
  emailHandler = this.inputHandler('email');
  passwordHandler = this.inputHandler('password');
  confirmPasswordHandler = this.inputHandler('confirmPassword');

  registerHandler = () => {
    this.setState(
      {
        touchedFields: {
          email: true,
          password: true,
          confirmPassword: true
        }
      },
      () => {
        const errors = validate(this.state);
        const areErrors = errObj =>
          Object.keys(errObj).map(key => errors[key]).some(Boolean);

        if (!areErrors(errors)) {
          this.props.register({
            email: this.state.email,
            password: this.state.password,
            confirmPassword: this.state.confirmPassword
          });
        }
      }
    );
  };

  render() {
    const errors = validate(this.state);

    return (
      <SessionContainer>
        <Title center>Register</Title>
        <FormField>
          <WithErrorInput
            key="email"
            error={errors.email}
            value={this.state.email}
            onChange={this.emailHandler}
            placeholder="Email"
            type="text"
          />
        </FormField>
        <FormField>
          <WithErrorInput
            key="password"
            error={errors.password}
            onChange={this.passwordHandler}
            value={this.state.password}
            placeholder="Enter password"
            type="password"
          />
        </FormField>
        <FormField>
          <WithErrorInput
            key="confirmPassword"
            error={errors.confirmPassword}
            onChange={this.confirmPasswordHandler}
            value={this.state.confirmPassword}
            placeholder="Retype password"
            type="password"
          />
        </FormField>
        <Button
          primary
          onClick={this.registerHandler}
          style={{ marginBottom: '20px' }}
        >
          Register
        </Button>
        <Link to="/signin">Signin</Link>
      </SessionContainer>
    );
  }
}
