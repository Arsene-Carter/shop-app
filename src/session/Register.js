import React, { Component } from 'react';
import { Link } from 'react-router';

export default function Register() {
    return (
        <div>
            <h2>Register</h2>
            <Link to='/'>Home</Link>
        </div>
    )
}