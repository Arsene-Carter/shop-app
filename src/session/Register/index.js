import React, { PropTypes, Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { Input, Button } from '../../ui';
import { RegisterContainer, RegisterTitle } from './RegisterStyles';
import * as actions from '../../actions';

class Register extends Component {
  inputHandler = name => e => this.setState({ [name]: e.target.value })
  emailHandler = this.inputHandler('email')
  passwordHandler = this.inputHandler('password')
  confirmPasswordHandler = this.inputHandler('confirmPassword')
  registerHandler = () => 
    this.props.register({
      email: this.state.email,
      password: this.state.password,
      confirmPassword: this.state.confirmPassword,
    })

  render() {
    return (
      <RegisterContainer>
        <RegisterTitle>Register</RegisterTitle>
        <Input
          onChange={this.emailHandler}
          placeholder="Enter email"
        />
        <Input
          onChange={this.passwordHandler}
          placeholder="Enter password"
          type="password"
        />
        <Input
          onChange={this.confirmPasswordHandler}
          placeholder="Retype password"
          type="password"
        />
        <Button
          onClick={this.registerHandler}
          styles={{ alignSelf: 'center', marginTop: '20px' }}
          primary
        >
          Register
        </Button>
      </RegisterContainer>
    );
  }
}

Register.propTypes = {
  register: PropTypes.func.isRequired,
};

export default connect(null, actions)(Register);
