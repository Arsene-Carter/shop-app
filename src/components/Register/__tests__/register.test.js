import renderer from 'react-test-renderer';
import React from 'react';
import { createBrowserHistory } from 'history';
import { MemoryRouter } from 'react-router-dom';
import Register, { validate } from '../';
import WithStyletron from '../../../shared/styletron/configure';
import * as validationErrors from '../../validationErrors';

it('render register', () => {
  const registerTree = renderer.create(
    <MemoryRouter>
      <WithStyletron>
        <Register history={createBrowserHistory()} />
      </WithStyletron>
    </MemoryRouter>
  ).toJSON();
  expect(registerTree).toMatchSnapshot();
});


test('register validate all fields are required', () => {
  const fields = {
    email: '',
    password: '',
    confirmPassword: '',
    touchedFields: {
      email: true,
      password: true,
      confirmPassword: true
    }
  }
  const requiredString = 'This field is required';

  expect(validate(fields)).toEqual({
    email: requiredString,
    password: requiredString,
    confirmPassword: requiredString
  })
})

test('register validate email', () => {
  const fields = {
    email: 'ioan',
    password: 'password',
    confirmPassword: 'password',
    touchedFields: {
      email: true,
      password: true,
      confirmPassword: true
    }
  }


  expect(validate(fields)).toMatchObject({
    email: validationErrors.INVALID_EMAIL
  })
})

test('register validate password length', () => {
  const fields = {
    email: 'ioan',
    password: 'pass',
    confirmPassword: 'pass',
    touchedFields: {
      email: true,
      password: true,
      confirmPassword: true
    }
  }


  expect(validate(fields)).toMatchObject({
    password: validationErrors.MIN_6_CHARACTERS
  })
})

test('if a field is not touched should not be validated', () => {
  const fields = {
    email: '',
    password: '',
    confirmPassword: '',
    touchedFields: {
      email: false,
      password: false,
      confirmPassword: false
    }
  }

  expect(validate(fields)).toMatchObject({
    email: '',
    password: '',
    confirmPassword: ''
  })
})

test('password should match confirm password', () => {
  const fields = {
    email: '',
    password: 'password',
    confirmPassword: 'password1',
    touchedFields: {
      email: false,
      password: true,
      confirmPassword: true
    }
  }

  expect(validate(fields)).toMatchObject({
    email: '',
    password: '',
    confirmPassword: validationErrors.PASSWORD_NOT_MATCH
  })
})
