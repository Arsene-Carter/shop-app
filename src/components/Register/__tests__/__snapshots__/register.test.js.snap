exports[`test render register 1`] = `
<div
  className="a b c d e f g h i j k l m n o p q r s t u v">
  <h2
    className="">
    Register
  </h2>
  <div
    className={undefined}>
    <div>
      <div
        className={undefined}>
        <input
          className=""
          onChange={[Function]}
          placeholder="Email"
          type="text"
          value="" />
      </div>
      
    </div>
  </div>
  <div
    className={undefined}>
    <div>
      <div
        className={undefined}>
        <input
          className=""
          onChange={[Function]}
          placeholder="Enter password"
          type="password"
          value="" />
      </div>
      
    </div>
  </div>
  <div
    className={undefined}>
    <div>
      <div
        className={undefined}>
        <input
          className=""
          onChange={[Function]}
          placeholder="Retype password"
          type="password"
          value="" />
      </div>
      
    </div>
  </div>
  <a
    className=""
    onClick={[Function]}
    style={
      Object {
        "marginBottom": "20px",
      }
    }>
    Register
  </a>
  <a
    className=""
    href="/signin"
    onClick={[Function]}>
    Signin
  </a>
</div>
`;
